### PR TITLE
Dependancy upgrades-update on pickle model-cython code updates-error-handling

### PR DESCRIPTION
**Pull request by tim-githaiga**

### CODE REVIEW

File: templates/index.html [hunk 1]
```diff
@@ -45,7 +45,7 @@ <h3 id="prediction"></h3>
     </div> 
 </div>
 <br>
-<p>©2021 VAIBHAV BICHAVE</p>
+<p>©2024 TIM</p>
```
Comment (vaibhavbichave) [inline templates/index.html:48]: Vaibhav Bichave

File: app.py [hunk 2]
```diff
@@ -9,31 +7,36 @@
...
-
+        try:
+            obj = FeatureExtraction(url)
+            features = obj.getFeaturesList()
+            x = np.array(features).reshape(1, -1)  # Adjust reshape if needed
+
+            if not hasattr(gbc, 'predict'):
+                raise ValueError("Loaded object is not a valid model.")
+            
+            y_pred = gbc.predict(x)[0]
```
Comment (vaibhavbichave) [inline app.py:31]: Add a comment for safe and unsafe

File: app.py [hunk 3]
```diff
@@ -9,31 +7,36 @@
...
-        y_pro_non_phishing = gbc.predict_proba(x)[0,1]
-        # if(y_pred ==1 ):
-        pred = "It is {0:.2f} % safe to go ".format(y_pro_phishing*100)
-        return render_template('index.html',xx =round(y_pro_non_phishing,2),url=url )
-    return render_template("index.html", xx =-1)
-
+        try:
+            obj = FeatureExtraction(url)
+            features = obj.getFeaturesList()
+            x = np.array(features).reshape(1, -1)  # Adjust reshape if needed
```
Comment (vaibhavbichave) [inline app.py:26]: x = np.array(obj.getFeaturesList()).reshape(1, -1)

File: trin.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
+# trin.py
```
Comment (vaibhavbichave) [inline trin.py:1]: File Name should be train.py ?

File: trin.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+def save_model(model, model_file='pickle/model.pkl'):
+    """Save the trained model to a file."""
+    joblib.dump(model, model_file)
+    print(f"Model saved as {model_file}")
+
+if __name__ == "__main__":
+    # Load data
+    csv_file = 'phishing.csv'
+    df = load_data(csv_file)
```
Comment (vaibhavbichave) [inline trin.py:40]: df = load_data('phishing.csv')

File: trin.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+def preprocess_data(df):
+    """Preprocess the dataset, separating features and target."""
+    X = df.drop(columns=['class'])  # Features
+    y = df['class']                 # Target
+    return X, y
+
+def train_model(X_train, y_train):
+    """Train a RandomForestClassifier model."""
+    model = RandomForestClassifier(n_estimators=100, random_state=42)
+    model.fit(X_train, y_train)
```
Comment (vaibhavbichave) [inline trin.py:23]: Can we use'Gradient Boosting Classifier' instead of 'RandomForestClassifier'
